examples/sst2/train.py

```diff
@@ -0,0 +1,272 @@
+# Copyright 2021 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Trains an SST2 text classifier."""
+import copy
+import functools
+from typing import Any, Dict, Iterable, Mapping, Optional, Sequence, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+import ml_collections
+import numpy as np
+from absl import logging
+from flax import linen as nn
+from flax import optim
+from flax.metrics import tensorboard
+import tensorflow as tf
+
+import input_pipeline
+import models
+
+Array = jnp.ndarray
+Example = Dict[str, Array]
+logging.set_verbosity(logging.DEBUG)
+
+
+@jax.vmap
+def sigmoid_cross_entropy_with_logits(*, labels: Array, logits: Array) -> Array:
+  """Sigmoid cross entropy loss."""
+  zeros = jnp.zeros_like(logits, dtype=logits.dtype)
+  condition = (logits >= zeros)
+  relu_logits = jnp.where(condition, logits, zeros)
+  neg_abs_logits = jnp.where(condition, -logits, logits)
+  return relu_logits - logits * labels + jnp.log1p(jnp.exp(neg_abs_logits))
+
+
+def get_initial_params_and_state(key, model):
+  """Returns randomly initialized parameters and a fresh model state."""
+  token_ids = jnp.ones((2, 3), jnp.int32)
+  lengths = jnp.ones((2,), dtype=jnp.int32)
+  variables = model.init(key, token_ids, lengths)
+  state, params = variables.pop('params')
+  return params, state
+
+
+def create_optimizer(params, learning_rate, beta, weight_decay):
+  """Returns a momentum optimizer."""
+  optimizer_def = optim.Momentum(
+      learning_rate=learning_rate,
+      beta=beta,
+      weight_decay=weight_decay)
+  optimizer = optimizer_def.create(params)
+  return optimizer
+
+
+def compute_metrics(*, labels: Array, logits: Array) -> Dict[str, Array]:
+  """Computes the metrics, summed across the batch if a batch is provided."""
+  if labels.ndim == 1:  # Prevent the labels from broadcasting over the logits.
+    labels = jnp.expand_dims(labels, axis=1)
+  loss = sigmoid_cross_entropy_with_logits(labels=labels, logits=logits)
+  binary_predictions = (logits >= 0.)
+  binary_accuracy = jnp.equal(binary_predictions, labels)
+  metrics = {
+      'loss': jnp.sum(loss),
+      'accuracy': jnp.sum(binary_accuracy),
+      'count': logits.shape[0]
+  }
+  return metrics
+
+
+@functools.partial(jax.jit, static_argnums=(0,))
+def train_step(
+    model: nn.Module,
```

Comment:
I think it is more common to construct the `Module` when you need it, rather than passing it around, since you then don't have to worry about static argnums.

So you could just pass `config` to this function, and then do `model_from_config(config).apply(...)` in `loss_fn` below.